### DISCUSSION TITLE
How to create river type dataset from pandas dataframe

### DISCUSSION THREAD

**Discussion by jacowp357** [online-ml/river#630] [Q&A] [answered]
I would like to see an example of creating a `datasets.Phishing()` type dataset from an arbitrary pandas dataframe? Is there such an example somewhere? This is for the purpose of using `evaluate.progressive_val_score()`

**Comment by raphaelsty** [comment, marked as answer]
Hi @jacowp357,

The classic structure of datasets with River is a list of dictionary. Here is how to convert a pandas dataframe to a list of dictionary:

```python
import pandas as pd

# Load Iris dataset.
df = pd.read_csv('https://raw.githubusercontent.com/mwaskom/seaborn-data/master/iris.csv')

dataset = df.to_dict(orient = "records")

dataset
```

```
[
  {'sepal_length': 5.1,
    'sepal_width': 3.5,
    'petal_length': 1.4,
    'petal_width': 0.2,
    'species': 'setosa'},
   {'sepal_length': 4.9,
    'sepal_width': 3.0,
    'petal_length': 1.4,
    'petal_width': 0.2,
    'species': 'setosa'},
...
]
```

You can pass the list of dict to `evaluate.progressive_val_score()` 🙂

**Reply by jacowp357** [reply]
Hi @raphaelsty,

When I do that with my dataframe I get this:

`[
{'x0': 15.684943000377492, 'x1': -1.6830765284127822, 'target': False},
 {'x0': 17.309895587880902, 'x1': 3.133674178365953, 'target': True},
 {'x0': 17.085973247837096, 'x1': 3.0402340466127598, 'target': True},
 {'x0': 14.477562519960992, 'x1': 2.101431188294361, 'target': True},
 {'x0': 14.999780445163736, 'x1': 1.983501742848379, 'target': True},
...
]`

Note the target variable is also in the dicts.

I get the following error when passing this to `evaluate.progressive_val_score()`

<img width="387" alt="Screenshot 2021-07-07 at 14 14 03" src="https://user-images.githubusercontent.com/16148279/124757185-9e762f80-df2d-11eb-87df-3f9e842bf35a.png">

**Reply by raphaelsty** [reply]
Of course, sorry, I answered too quickly without thinking. We have to create a list of tuples. The set of features is defined within a dictionary and the target variable is defined outside the dictionary.

```python
import pandas as pd

# Load Iris dataset.
df = pd.read_csv('https://raw.githubusercontent.com/mwaskom/seaborn-data/master/iris.csv')

# "target" instead of "species" for your dataset
dataset = [(x, x.pop("species")) for x in  df.to_dict(orient = "records")]
```

**Reply by jacowp357** [reply]
@raphaelsty thanks! That did the trick!

**Reply by MaxHalford** [reply]
We also have a `stream.iter_pandas` method :)

**Reply by Divya216** [reply]
Explain with an example stream.iter_pandas

**Reply by MaxHalford** [reply]
https://riverml.xyz/0.14.0/api/stream/iter-pandas/#examples
